### PR TITLE
[IEXPLORE][IEFRAME][SHELL32] Add Internet icon on Desktop

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

JIRA issue: [CORE-18625](https://jira.reactos.org/browse/CORE-18625)
relies on https://github.com/reactos/reactos/commit/1c91b0b61ddb893e7fb52e84c8d415fe426915de to work properly

## Proposed changes

- Add Internet icon on Desktop.
- Modify `"HKCR\CLSID\%CLSID_Internet%"` registry key.
- Add `IDS_INTERNET` and `IDS_INTERNET_DESCRITION` resource strings into `ieframe.dll`.
- Modify `dll/win32/shell32/folders/CDesktopFolder.cpp` and `dll/win32/shell32/folders/CRegFolder.cpp` to add the icon.

## Comparison

BEFORE:
before
No internet icon.

AFTER:
after
There is the Internet icon. You can hide/show the icon from Desktop's customization.

## TODO

- [x] Do tests.

### CODE REVIEW

**Comment by learn-more** [comment]
What happens when clicking this icon?

**Comment by katahiromz** [comment]
> What happens when clicking this icon?

Currently not working.

File: base/applications/iexplore/iexplore.inf
Comment (learn-more) [inline base/applications/iexplore/iexplore.inf:19]: I would comment these out as long as it is not working, that saves us from receiving 30 tickets that it does not work.

Comment (katahiromz) [reply base/applications/iexplore/iexplore.inf]: Commented out. But `shell32` keeps the icon shown.

VirtualBox_ReactOS_18_07_2023_17_17_45

File: dll/win32/ieframe/lang/cs-CZ.rc
Comment (HBelusca) [inline dll/win32/ieframe/lang/cs-CZ.rc:65]: lol

File: dll/win32/ieframe/lang/cs-CZ.rc
Comment (HBelusca) [inline dll/win32/ieframe/lang/cs-CZ.rc:64]: If this thing opens iexplore.exe unconditionally, I think this should be named `"Wine Internet Explrer"` to really insist that it's going to open Wine IE and not anything else (and that this program really comes from Wine, and not ReactOS.)

Comment (binarymaster) [reply dll/win32/ieframe/lang/cs-CZ.rc]: In Windows XP/2003, when modern start menu is enabled, there is "Internet" menu item, which starts the web browser assigned by the user - it's not necessarily "Internet Explorer":

image

File: dll/win32/shell32/lang/en-US.rc
Comment (binarymaster) [inline dll/win32/shell32/lang/en-US.rc:985]: "Internet Browser" please, to be consistent with desk.cpl setting:
```suggestion
    IDS_INTERNET "Internet Browser"
```

File: base/applications/iexplore/iexplore.inf
Comment (binarymaster) [inline base/applications/iexplore/iexplore.inf:25]: Maybe use an URL scheme here, so it would start default assigned browser instead. I know Windows XP/2003 opens Internet Explorer by clicking on this, but suggested behavior will be more useful.

File: base/applications/iexplore/iexplore.inf
Comment (binarymaster) [inline base/applications/iexplore/iexplore.inf:22]: Ideally this should be the globe icon `iexplore.exe,13`, but Wine IE has only one icon. Closest one I can find is shell32 `IDI_SHELL_ENTIRE_NETWORK`:
```suggestion
HKCR,"CLSID\%CLSID_Internet%\DefaultIcon",,0x00020000,"%SystemRoot%\system32\shell32.dll,-14"
```

File: dll/win32/shell32/folders/CRegFolder.cpp
Comment (binarymaster) [inline dll/win32/shell32/folders/CRegFolder.cpp:200]:
```suggestion
            ExpandEnvironmentStringsW(L"%SystemRoot%\\system32\\shell32.dll",
                                      szPath, _countof(szPath));
            initIcon->SetNormalIcon(szPath, -14);
```

File: base/applications/iexplore/iexplore.inf
Comment (katahiromz) [reply base/applications/iexplore/iexplore.inf]: Double-clicking action is fixed.

File: dll/win32/url/url_main.c
Comment (HBelusca) [inline dll/win32/url/url_main.c:110]: I would suggest you check what does Windows' url.dll does there, and whether it uses instead the
`void WINAPI OpenURL(HWND hWnd, HINSTANCE hInst, LPCSTR lpcstrUrl, int nShowCmd)` from ieframe.dll .... or whether it is the other way around.

File: dll/win32/shell32/wine/classes.c
Comment (HBelusca) [inline dll/win32/shell32/wine/classes.c:411]: Shouldn't this mapping come from the registry directly? :thinking: (I have no idea about that tbh)

File: dll/win32/shell32/lang/en-US.rc
Comment (HBelusca) [inline dll/win32/shell32/lang/en-US.rc:986]: For other translations as well:
```suggestion

    IDS_INTERNET "Internet Browser"
    IDS_INTERNET_DESCRIPTION "Opens a Web browser and displays information on the Internet."
```

File: dll/win32/url/url_main.c
Comment (katahiromz) [reply dll/win32/url/url_main.c]: image
Just calling `ShellExecuteA`. `OpenURL` is an alias of `OpenURLA`.

File: dll/win32/shell32/lang/en-US.rc
Comment (katahiromz) [reply dll/win32/shell32/lang/en-US.rc]: Fixed.

Comment (binarymaster) [reply dll/win32/shell32/lang/en-US.rc]: You did not fixed "Open" => "Opens"

File: dll/win32/shell32/wine/classes.c
Comment (katahiromz) [reply dll/win32/shell32/wine/classes.c]: According to my `MessageBox` debugging, it was needless code.

Deleted.

File: base/applications/iexplore/iexplore.inf
Comment (HBelusca) [inline base/applications/iexplore/iexplore.inf:25]: Well... why not another search engine? It would actually be better to just use the `about::blank` default page, so that there is no imposed choice of the search engine on the user.

Comment (binarymaster) [reply base/applications/iexplore/iexplore.inf:25]: Ideally it should open the homepage set in the Internet control panel options.

Comment (katahiromz) [reply base/applications/iexplore/iexplore.inf:25]: `https://` (only this) is preferred but ReactOS has a bug in opening `https://` (only this):

https://jira.reactos.org/browse/CORE-19042

**Comment by katahiromz** [comment]
We have to implement `CLSID_Internet` in `ieframe.dll` in the future...
https://git.reactos.org/?p=reactos.git;a=blob;f=modules/rostests/apitests/com/ieframe.c;hb=bf2cec186cc7655e062ec0e53ccfac860bcae70d#l35

File: base/applications/iexplore/iexplore.inf
Comment (katahiromz) [reply base/applications/iexplore/iexplore.inf:25]: `about:blank` is not applicable in this case.

**Review by binarymaster — COMMENTED** [review]
Some suggested comments and well... Wine

**Comment by JoachimHenze** [comment]
Question: Is it inevitable to add the 2 IDS_INTERNET "Internet Browser" & IDS_INTERNET_DESCRIPTION "Opens a Web browser and displays information on the Internet." within the explorers resources? Or couldn't they also be implemented within the shortcut.inf maybe?

File: dll/win32/url/url_main.c
Comment (katahiromz) [reply dll/win32/url/url_main.c]: Reverted and added `#ifdef __REACTOS__`.

**Comment by katahiromz** [comment]
> Is it inevitable to add the 2...

We can move them to `ieframe.dll`.

**Comment by JoachimHenze** [comment]
Afaik ieframe.dll isn't always loaded into memory. Given that would not change by the moving of the IDS, I would vote to move those IDS from shell32.dll to ieframe.dll.

File: base/applications/iexplore/iexplore.inf
Comment (JoachimHenze) [reply base/applications/iexplore/iexplore.inf]: I absolutely love the proposal with about:blank. Why isn't it applicable in this case? I do not understand yet.

Comment (katahiromz) [reply base/applications/iexplore/iexplore.inf]: `about:blank` is not a valid URL.

Comment (JoachimHenze) [reply base/applications/iexplore/iexplore.inf]: I can confirm that I cannot open
`rundll32.exe url, OpenURL about:blank` from the rundlg of XPSP3 with the aim to open IE.
But `rundll32.exe url, OpenURL about:blank https://google.com` works.
So maybe this is enough justification. Thx.
A pity, I would have liked an empty browser tab very much.
I would vote then for either "the homepage set in the Internet control panel options" or `https://google.com`
I do think both is more helpful here for the user than `https://reactos.org`

File: base/applications/iexplore/iexplore.inf
Comment (JoachimHenze) [inline base/applications/iexplore/iexplore.inf:28]: Omit the trailing slash please at least.
```suggestion
HKCR,"CLSID\%CLSID_Internet%\Shell\open\Command",,,"rundll32.exe url,OpenURL https://reactos.org"
```

**Review by JoachimHenze — DISMISSED** [review]
Omit the trailing slash at least,
and if it wouldn't impose ieframe.dll to be loaded in memory all the time, then please move the IDS over there.

File: dll/win32/ieframe/ieframe_v1.rgs
Comment (JoachimHenze) [inline dll/win32/ieframe/ieframe_v1.rgs:8]: why that change?

Comment (binarymaster) [reply dll/win32/ieframe/ieframe_v1.rgs:8]: This is for consistency with desk.cpl option.

Comment (katahiromz) [reply dll/win32/ieframe/ieframe_v1.rgs:8]: The international settings are not correctly applied on Desktop icon text.

Comment (JoachimHenze) [reply dll/win32/ieframe/ieframe_v1.rgs:8]: We could also use just "Internet" in the dek.cpl to harmonize. Would even make the translation easier, and "Internet" is more an internationalism than "Internet Browser" imho. So it may avoid the need for translation for some languages entirely. I would volunteer personally to harmonize desk.cpl later.

Comment (binarymaster) [reply dll/win32/ieframe/ieframe_v1.rgs:8]: Windows desk.cpl uses two words "Internet Explorer", using one word there wouldn't look consistent. I would not prefer that UX decision.

Comment (JoachimHenze) [reply dll/win32/ieframe/ieframe_v1.rgs:8]: That is no argument at all, because "Internet Browser" is also not the same as "Internet Explorer".

Comment (binarymaster) [reply dll/win32/ieframe/ieframe_v1.rgs:8]: There are 2 words in both cases, it looks consistent from the UI point of view. Using just one word there would make it an oddball. Also using just "Internet" is quite ambiguous. These are two reasons I have chosen "Internet Browser" when I worked on desktop icons settings in desk.cpl.

Comment (JoachimHenze) [reply dll/win32/ieframe/ieframe_v1.rgs:8]: Let's see whether there are more upvotes for your or for my position ultimately and then stick with the decision of the majority. Let the thumbs decide. My current rejection of this PR is not due to this specific discussion-thread.

File: dll/win32/url/url_main.c
Comment (JoachimHenze) [inline dll/win32/url/url_main.c:113]: It might be worth to split off the changes to `dll/win32/url/url_main.c` into a separate PR, as I think that part is much less controversial than the rest of this PR. It might be possible to get approval for that part much easier than for the rest of this PR. E.g. Personally I would approve the url_main part right away.
The rest of this PR may be a bit controversial for some of us because it couples the mshtml/ie part more tightly to the shell, which is something many of the devs tried to avoid and commonly agreed to, some years ago.

File: base/applications/iexplore/iexplore.rc
Comment (JoachimHenze) [inline base/applications/iexplore/iexplore.rc:30]: Does adding the quotations fix anything specific?

File: base/applications/iexplore/iexplore.rc
Comment (JoachimHenze) [inline base/applications/iexplore/iexplore.rc:33]: Does adding the quotations fix anything specific?

Comment (katahiromz) [reply base/applications/iexplore/iexplore.rc]: Reverted.

File: base/applications/iexplore/iexplore.rc
Comment (katahiromz) [reply base/applications/iexplore/iexplore.rc]: Reverted.

File: dll/win32/url/url_main.c
Comment (katahiromz) [reply dll/win32/url/url_main.c:113]: Fixed in `master`.

File: base/applications/iexplore/iexplore.inf
Comment (katahiromz) [reply base/applications/iexplore/iexplore.inf]: Used `https://www.google.com`.

File: base/applications/iexplore/iexplore.inf
Comment (katahiromz) [reply base/applications/iexplore/iexplore.inf]: Used `https://www.google.com`.

**Comment by JoachimHenze** [comment]
All my most important change requests have been addressed properly by Katayama. I changed back to neutrality for this PR therefore. Will not approve yet due to the few remaining unresolved discussion threads.

File: dll/win32/url/url_main.c
Comment (JoachimHenze) [reply dll/win32/url/url_main.c:113]: Can be stripped now in here, because got merged already.

Comment (katahiromz) [reply dll/win32/url/url_main.c]: Deleted change.

File: dll/win32/ieframe/ieframe_v1.rgs
Comment (katahiromz) [reply dll/win32/ieframe/ieframe_v1.rgs:8]: https://chat.reactos.org/reactos/pl/ysjmbfeempff9pwd3ofwbzaz8r

Comment (katahiromz) [reply dll/win32/ieframe/ieframe_v1.rgs:8]: Voting: https://chat.reactos.org/reactos/pl/woka83o84jbnuy1m33r7kqeo4r

File: base/applications/iexplore/iexplore.inf
Comment (HBelusca) [reply base/applications/iexplore/iexplore.inf]: Not everyone is fond of google by the way (e.g. in France).
https://alternativeto.net/software/google-search/?license=opensource

File: dll/win32/ieframe/ieframe_v1.rgs
Comment (katahiromz) [reply dll/win32/ieframe/ieframe_v1.rgs:8]: Charged to `Internet`.

Comment (binarymaster) [reply dll/win32/ieframe/ieframe_v1.rgs:8]: The poll in the chat has been started in the time when I was not online, and the change has been pushed earlier than 24 hours from the poll start, this is illegal.

File: base/applications/iexplore/iexplore.inf
Comment (erkinalp) [inline base/applications/iexplore/iexplore.inf:28]:
```suggestion
HKCR,"CLSID\%CLSID_Internet%\Shell\open\Command",,,"rundll32.exe url,OpenURL https://www.reactos.org"
```

**Review by erkinalp — CHANGES_REQUESTED** [review]
fixed the trace of el Goog

File: base/applications/iexplore/iexplore.inf
Comment (erkinalp) [reply base/applications/iexplore/iexplore.inf]: I have proposed a revert below...

Comment (erkinalp) [reply base/applications/iexplore/iexplore.inf]: @JoachimHenze about:blank is not an URI, but a browser magic string.

File: base/applications/iexplore/iexplore.inf
Comment (binarymaster) [inline base/applications/iexplore/iexplore.inf:24]:
```suggestion
HKCR,"CLSID\%CLSID_Internet%\DefaultIcon",,0x00020000,"%SystemRoot%\system32\shell32.dll,-512"
```

File: dll/win32/shell32/folders/CRegFolder.cpp
Comment (binarymaster) [inline dll/win32/shell32/folders/CRegFolder.cpp:196]:
```suggestion
            initIcon->SetNormalIcon(swShell32Name, -IDI_SHELL_WEB_BROWSER);
```

**Review by binarymaster — COMMENTED** [review]
Found a better fitting icon: `shell32.dll,-512` (`IDI_SHELL_WEB_BROWSER`)

File: base/applications/iexplore/iexplore.inf
Comment (katahiromz) [reply base/applications/iexplore/iexplore.inf:28]: I don't agree.

File: dll/win32/ieframe/ieframe_v1.rgs
Comment (katahiromz) [reply dll/win32/ieframe/ieframe_v1.rgs:8]: You can start a poll it again if you want.

Comment (katahiromz) [reply dll/win32/ieframe/ieframe_v1.rgs:8]: Reverted to `"Internet Browser"`.

File: base/applications/iexplore/iexplore.inf
Comment (JoachimHenze) [inline base/applications/iexplore/iexplore.inf:28]:
```suggestion
HKCR,"CLSID\%CLSID_Internet%\Shell\open\Command",,,"rundll32.exe url,OpenURL https://google.com"
```

File: base/applications/iexplore/iexplore.inf
Comment (binarymaster) [inline base/applications/iexplore/iexplore.inf:30]:
```suggestion
HKCR,"CLSID\%CLSID_Internet%\ShellFolder",,2,"0x24"

```

**Review by binarymaster — APPROVED** [review]
One minor suggestion to reduce the diff, and then it's good to go. :+1:

**Pull request merged**
